### DISCUSSION TITLE
Using Suspense with async write action

### DISCUSSION THREAD

**Discussion by janrembold** [pmndrs/jotai#1195] [Q&A] [answered]
Hi , I'm currently playing around with Jotai and was wondering why `Suspense` worked fine with a simple async atom but not with an async fetch action. 

I wrote a simple Codesandbox: https://codesandbox.io/s/jotai-async-fetch-suspense-35j93v?file=/src/App.js

`Component1` uses `useEffect` to call 4s delayed `fetchData`. Looks like `Suspense` was not called or at least not long enough to be visible. I guess because there is already a (default) value available in `dataAtom`.

What would be the best practice to use some async action together with `Suspense`?

Thanks in advance for any help!
Best, Jan

**Comment by dai-shi** [comment, marked as answer]
Thanks for asking.
Well, that's what it is: async `read` suspends, but async `write` does not.
Historically, I thought too that async write suspends, and implemented so.
But that was causing too many troubles and noticed a bad design. So, changed in https://github.com/pmndrs/jotai/releases/tag/v1.4.0 and #731.

Now, what should we do if we want to suspend using `write`?
We can store a promise value.
Check this out: https://codesandbox.io/s/jotai-async-fetch-suspense-forked-sf2ckk

#1176 is to improve docs around this.

**Reply by janrembold** [reply]
Thanks for the quick response!!! This is really impressive what you are building up here!

The proposed answer works as expected in JS but in TS I see a typing error for `set(dataAtom, promise)` that says `The argument of type "Promise<string>" cannot be assigned to the parameter of type "SetStateAction<string>". ts(2345)`

I can simply set `promise as unknown as SetStateAction<string>` but this is a bad workaround. Shouldn't those types be pretty much the same? What is the proposed typing here?

**Reply by dai-shi** [reply]
You can define the atom with annotation: `atom<string | Promise<string>>('initial value')`

**Reply by janrembold** [reply]
So easy!!! Thanks mate!

**Reply by HJordan35** [reply]
I know this is an old thread - but wanted to see if there were any recommendations to remove the flicker caused by setting that initial value when using the `write` async atom approach. Ive been attempting this with both `loadable` and `suspense` but am experiencing a brief moment where that initial value is interpreted before the `setter` can update the initial value to a promise

- when using `loadable` you get a situation where state is set to `hasData` due to the initial value
- when using `suspense` the app will flicker and show the content

The only thoughts I have to resolve so far would be either set a promise as the initial value with some sort of arbitrary   timeout (but that feels kind of wrong) - or creating my own loading state system - but would love to hear any thoughts

**Reply by dai-shi** [reply]
> setting that initial value when using the write async atom approach.

I'm not sure if I understand the situation. Can you share a small code snippet?
Maybe it's better to open a new discussion.
We have a new util called `unwrap`, but I don't think it's related in this case.
